src/smart-components/Thread/components/ThreadList/ThreadListItem.tsx

```diff
@@ -0,0 +1,266 @@
+import React, { useMemo, useState, useRef, useEffect, useLayoutEffect } from 'react';
+import format from 'date-fns/format';
+import { FileMessage, UserMessage } from '@sendbird/chat/message';
+
+import { useLocalization } from '../../../../lib/LocalizationContext';
+import DateSeparator from '../../../../ui/DateSeparator';
+import Label, { LabelTypography, LabelColors } from '../../../../ui/Label';
+import RemoveMessage from '../RemoveMessageModal';
+import FileViewer from '../../../../ui/FileViewer';
+import { useThreadContext } from '../../context/ThreadProvider';
+import MessageContent from '../../../../ui/MessageContent';
+import useSendbirdStateContext from '../../../../hooks/useSendbirdStateContext';
+import SuggestedMentionList from '../../../Channel/components/SuggestedMentionList';
+import MessageInput from '../../../../ui/MessageInput';
+import { ThreadListStateTypes } from '../../types';
+import { MessageInputKeys } from '../../../../ui/MessageInput/const';
+
+export interface ThreadListItemProps {
+  className?: string;
+  message: UserMessage | FileMessage;
+  chainTop?: boolean;
+  chainBottom?: boolean;
+  hasSeparator?: boolean;
+  renderCustomSeparator?: (props: { message: UserMessage | FileMessage }) => React.ReactElement;
+  handleScroll?: () => void;
+}
+
+export default function ThreadListItem({
+  className,
+  message,
+  chainTop,
+  chainBottom,
+  hasSeparator,
+  renderCustomSeparator,
+  handleScroll,
+}: ThreadListItemProps): React.ReactElement {
+  const { stores, config } = useSendbirdStateContext();
+  const {
+    isReactionEnabled,
+    isMentionEnabled,
+    isOnline,
+    replyType,
+    userMention,
+  } = config;
+  const userId = stores?.userStore?.user?.userId;
+  const { dateLocale } = useLocalization();
+  const {
+    currentChannel,
+    nicknamesMap,
+    emojiContainer,
+    toggleReaction,
+    threadListStatus,
+    updateMessage,
+    isMuted,
+    isChannelFrozen,
+  } = useThreadContext();
+
+  const [showEdit, setShowEdit] = useState(false);
+  const [showRemove, setShowRemove] = useState(false);
+  const [showFileViewer, setShowFileViewer] = useState(false);
+  const usingReaction = isReactionEnabled && !currentChannel?.isSuper && !currentChannel?.isBroadcast;
+
+  // reactions
+  useLayoutEffect(() => {
+    handleScroll?.();
+  }, [showEdit, message?.reactions?.length]);
+
+  // mention
+  const editMessageInputRef = useRef(null);
+  const [mentionNickname, setMentionNickname] = useState('');
+  const [mentionedUsers, setMentionedUsers] = useState([]);
+  const [mentionedUserIds, setMentionedUserIds] = useState([]);
+  const [messageInputEvent, setMessageInputEvent] = useState(null);
+  const [selectedUser, setSelectedUser] = useState(null);
+  const [mentionSuggestedUsers, setMentionSuggestedUsers] = useState([]);
+  const [ableMention, setAbleMention] = useState(true);
+  const displaySuggestedMentionList = isOnline
```

Comment:
we should probably reorganize similarly scattered states back into the useReducer in parent Provider

Reply:
Yea I have to